File: backend/app/schemas/ripple.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from uuid import UUID
from app.schemas.user import UserBrief


class RippleCreateRequest(BaseModel):
    comment: Optional[str] = None


class RipplePushResponse(BaseModel):
    id: UUID
    target_user: UserBrief
    status: str
    delivered_at: Optional[datetime] = None
    viewed_at: Optional[datetime] = None

    model_config = {"from_attributes": True}


class RippleResponse(BaseModel):
    id: UUID
    skill_id: UUID
    skill_name: str
    skill_display_name: str
    sender: UserBrief
    pushes: List[RipplePushResponse] = []
    created_at: datetime

    model_config = {"from_attributes": True}


class RippleNotification(BaseModel):
    type: str = "ripple"
    ripple_id: UUID
    skill_name: str
    skill_slug: str
    sender: UserBrief
    comment: Optional[str] = None


class UpdateNotification(BaseModel):
    type: str = "skill_update"
    skill_name: str
    skill_slug: str
    new_version: str
    changelog: Optional[str] = None

File: packages/contract/src/ripple.ts
import { z } from 'zod';
import { userBriefSchema } from './user.js';

export const pushStatusSchema = z.enum(['pending', 'shown', 'consumed', 'dismissed']);
export type PushStatus = z.infer<typeof pushStatusSchema>;

export const ripplePushSchema = z.object({
  id: z.string().uuid(),
  target_user: userBriefSchema.nullable(),
  status: pushStatusSchema,
  shown_at: z.string().nullable(),
  consumed_at: z.string().nullable(),
});
export type RipplePush = z.infer<typeof ripplePushSchema>;

export const rippleRecordSchema = z.object({
  id: z.string().uuid(),
  skill_id: z.string().uuid(),
  skill_name: z.string(),
  skill_display_name: z.string(),
  sender: userBriefSchema,
  comment: z.string().nullable(),
  pushes: z.array(ripplePushSchema),
  created_at: z.string(),
});
export type RippleRecord = z.infer<typeof rippleRecordSchema>;

// ---- SSE 通知 ----

export const rippleNotificationSchema = z.object({
  type: z.literal('ripple'),
  delivery_id: z.string().uuid(),
  ripple_id: z.string().uuid(),
  skill_name: z.string(),
  skill_display_name: z.string(),
  skill_slug: z.string(),
  sender: userBriefSchema,
  comment: z.string().optional(),
});
export type RippleNotification = z.infer<typeof rippleNotificationSchema>;

export const updateNotificationSchema = z.object({
  type: z.literal('skill_update'),
  skill_name: z.string(),
  skill_display_name: z.string(),
  skill_slug: z.string(),
  new_version: z.string(),
  changelog: z.string().optional(),
});
export type UpdateNotification = z.infer<typeof updateNotificationSchema>;

export const notificationSchema = z.discriminatedUnion('type', [
  rippleNotificationSchema,
  updateNotificationSchema,
]);
export type Notification = z.infer<typeof notificationSchema>;

export const guestTouchResponseSchema = z.object({
  session_key: z.string(),
  active: z.boolean(),
});
export type GuestTouchResponse = z.infer<typeof guestTouchResponseSchema>;
